File: frontend/src/api/graph.ts
/** 知识图谱模块 REST 客户端（对齐后端 schemas/graph.py / api接口清单 §四）。 */
import { api } from "./client";

export type CallDirection = "BOTH" | "CALLERS" | "CALLEES";
export type Granularity = "MODULE" | "PACKAGE" | "CLASS";
export type NodeKind = "class" | "method" | "doc";

export interface GraphNode {
  id: string;
  name: string;
  type: string; // method/class/block/file/code/doc/module/package
  module?: string | null;
  class_name?: string | null;
  method_name?: string | null;
  file_path?: string | null;
  heading_path?: string[];
  stale?: boolean;
  stale_reason?: string | null;
  class_count?: number | null;
  depth?: number | null;
}

export interface GraphEdge {
  source: string;
  target: string;
  type: string; // CALLS / DOCUMENTED_BY / DEPENDS_ON
  weight: number;
  stale?: boolean;
  stale_reason?: string | null;
}

export interface GraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
  center?: string | null;
  truncated?: boolean;
}

export interface GraphSearchItem {
  id: string;
  name: string;
  type: string;
  module?: string | null;
  class_name?: string | null;
  file_path?: string | null;
  heading_path?: string[];
}

export interface GraphSearchResponse {
  items: GraphSearchItem[];
}

// ---- 调用图 ----
export const getCallGraph = (params: {
  center_node: string;
  depth?: number;
  direction?: CallDirection;
  max_nodes?: number;
}) => api.get<GraphResponse>("/v1/graph/call-graph", { params }).then((r) => r.data);

// ---- 代码-文档关联图 ----
export const getCodeDocRelations = (params: {
  center_node: string;
  depth?: number;
  include_stale_only?: boolean;
  max_nodes?: number;
}) => api.get<GraphResponse>("/v1/graph/code-doc-relations", { params }).then((r) => r.data);

// ---- 模块依赖图 ----
export const getModuleDependency = (params: { granularity?: Granularity } = {}) =>
  api.get<GraphResponse>("/v1/graph/module-dependency", { params }).then((r) => r.data);

// ---- 图谱节点搜索 ----
export const searchGraphNodes = (params: { q: string; node_type?: NodeKind; limit?: number }) =>
  api.get<GraphSearchResponse>("/v1/graph/search", { params }).then((r) => r.data);

File: backend/app/schemas/graph.py
"""知识图谱模块请求/响应 schema（Phase 4，对齐 api接口清单 §四）。

图节点/边为通用结构，由 4 个端点复用：
- ``call-graph``：code 节点（method/class）+ CALLS 边
- ``code-doc-relations``：code + doc 节点 + DOCUMENTED_BY 边（含 stale 标记）
- ``module-dependency``：module/package/class 聚合节点 + DEPENDS_ON 边（weight=跨组调用边数）
- ``search``：图谱节点搜索（返回可作 center_node 的 id）
"""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class GraphNode(BaseModel):
    """图节点。``type`` ∈ method/class/code/doc/module/package。"""
    id: str
    name: str
    type: str
    module: str | None = None
    class_name: str | None = None
    method_name: str | None = None
    file_path: str | None = None
    heading_path: list[str] = []
    stale: bool = False
    stale_reason: str | None = None
    class_count: int | None = None  # module/package 节点：组内不同 class 数
    depth: int | None = None  # 距 center 的 BFS 跳数（call-graph / code-doc）


class GraphEdge(BaseModel):
    """图边。``type`` ∈ CALLS / DOCUMENTED_BY / DEPENDS_ON。"""
    source: str
    target: str
    type: str
    weight: int = 1
    stale: bool = False
    stale_reason: str | None = None


class GraphResponse(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    center: str | None = None
    truncated: bool = False  # 达到 max_nodes 被截断


class GraphSearchItem(BaseModel):
    id: str  # method/doc → chunk_id；class → ``class:{ClassName}``
    name: str
    type: str  # class / method / doc
    module: str | None = None
    class_name: str | None = None
    file_path: str | None = None
    heading_path: list[str] = []


class GraphSearchResponse(BaseModel):
    items: list[GraphSearchItem]


# ---- 查询参数枚举（文档化，路由侧用 Literal 校验） ----

CallDirection = Literal["BOTH", "CALLERS", "CALLEES"]
Granularity = Literal["MODULE", "PACKAGE", "CLASS"]
NodeKind = Literal["class", "method", "doc"]
